### PR TITLE
8283469: Don't use memset to initialize members in FileMapInfo and fix memory leak

### PR DESCRIPTION
**Pull request by zhengyu123**
`FileMapInfo` uses `memset` in its constructor to initialize data member, it is generally a bad idea. It can break things badly if adds a new member with a non-trivial constructor and `memset` is not suitable.

Besides, it leaks malloc'd `_header` member.

Test:

- [x]  hotspot_cds
- [x] tier1
- [x] tier2
- [x] runtime/cds/DeterministicDump.java

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283469](https://bugs.openjdk.java.net/browse/JDK-8283469): Don't use memset to initialize members in FileMapInfo and fix memory leak

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7899/head:pull/7899` \
`$ git checkout pull/7899`

Update a local copy of the PR: \
`$ git checkout pull/7899` \
`$ git pull https://git.openjdk.java.net/jdk pull/7899/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7899`

View PR using the GUI difftool: \
`$ git pr show -t 7899`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7899.diff">https://git.openjdk.java.net/jdk/pull/7899.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
Isn't this one of the things we have to guarantee was zeroed at initialization to produce deterministic shared archive contents?

David

**Comment by iklam** [comment]
> Isn't this one of the things we have to guarantee was zeroed at initialization to produce deterministic shared archive contents?
> 
> David

`FileMapInfo` is derived from `CHeapObj<mtInternal>` so it's allocated by os::malloc, which zeros the memory when `DumpSharedSpaces == true`.

**Review by iklam — APPROVED** [review]
LGTM. Please run all CDS tests, including DeterministicDump.java.

**Comment by zhengyu123** [comment]
Friendly ping! Might I get a second review.

Thanks

**Review by kimbarrett — APPROVED** [review]
Looks good.

memxxx of an object is only correct under specific circumstances.  A
case where it's not is when the object has virtual functions, which
FileMapInfo does in debug builds, because CHeapObj derives from
AllocatedObj in debug builds and that class has virtual functions.
Clearing the entire object is zeroing out vtable info, and trying to
call those functions will almost certainly crash.  So the current code
is already invalid; there's no need to wait for future members with
non-trivial constructors.

It looks like FileMapInfo might really a union of two cases, the
static and !static case, with some members associated with only one or
the other case.  In the future it might be worth considering whether
that could be made more formal.

**Comment by zhengyu123** [comment]
Thanks for the reviews, @iklam @kimbarrett

**Pull request closed without merging**
